smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,290 @@
+]
+_LaunchConfigType: TypeAlias = (
+    "_LauncherAdapter[ExecutableProtocol, _EnvironMappingType]"
+)
+_UnkownType: TypeAlias = t.NoReturn
+
+
+@t.final
+class Dispatcher:
+    """A class capable of deciding which launcher type should be used to launch
+    a given settings type.
+    """
+
+    def __init__(
+        self,
+        *,
+        dispatch_registry: (
+            t.Mapping[type[LaunchArguments], _DispatchRegistration[t.Any, t.Any]] | None
+        ) = None,
+    ) -> None:
+        self._dispatch_registry = (
+            dict(dispatch_registry) if dispatch_registry is not None else {}
+        )
+
+    def copy(self) -> Self:
+        """Create a shallow copy of the Dispatcher"""
+        return type(self)(dispatch_registry=self._dispatch_registry)
+
+    @t.overload
+    def dispatch(
+        self,
+        args: None = ...,
+        *,
+        with_format: _FormatterType[_DispatchableT, _LaunchableT],
+        to_launcher: type[LauncherProtocol[_LaunchableT]],
+        allow_overwrite: bool = ...,
+    ) -> t.Callable[[type[_DispatchableT]], type[_DispatchableT]]: ...
+    @t.overload
+    def dispatch(
+        self,
+        args: type[_DispatchableT],
+        *,
+        with_format: _FormatterType[_DispatchableT, _LaunchableT],
+        to_launcher: type[LauncherProtocol[_LaunchableT]],
+        allow_overwrite: bool = ...,
+    ) -> None: ...
+    def dispatch(
+        self,
+        args: type[_DispatchableT] | None = None,
+        *,
+        with_format: _FormatterType[_DispatchableT, _LaunchableT],
+        to_launcher: type[LauncherProtocol[_LaunchableT]],
+        allow_overwrite: bool = False,
+    ) -> t.Callable[[type[_DispatchableT]], type[_DispatchableT]] | None:
+        """A type safe way to add a mapping of settings type to launcher type
+        to handle a settings instance at launch time.
+        """
+        err_msg: str | None = None
+        if getattr(to_launcher, "_is_protocol", False):
+            err_msg = f"Cannot dispatch to protocol class `{to_launcher.__name__}`"
+        elif getattr(to_launcher, "__abstractmethods__", frozenset()):
+            err_msg = f"Cannot dispatch to abstract class `{to_launcher.__name__}`"
+        if err_msg is not None:
+            raise TypeError(err_msg)
+
+        def register(args_: type[_DispatchableT], /) -> type[_DispatchableT]:
+            if args_ in self._dispatch_registry and not allow_overwrite:
+                launcher_type = self._dispatch_registry[args_].launcher_type
+                raise TypeError(
+                    f"{args_.__name__} has already been registered to be "
+                    f"launched with {launcher_type}"
+                )
+            self._dispatch_registry[args_] = _DispatchRegistration(
+                with_format, to_launcher
+            )
+            return args_
+
+        if args is not None:
+            register(args)
+            return None
+        return register
+
+    def get_dispatch(
```

Comment:
Why accept two different types for args when it will always be of type LaunchArguments?

Reply:
mostly just convenience! that way if I have
```py
args: LaunchArguments = ...
```
I can call
```py
dispatcher.get_dispatch(args)
```
instead of
```py
dispatcher.get_dispatch(type(args))
```
If you think this is more confusing than it is helpful, I'd be willing to remove it!